internal/cephfs/store/volumeoptions.go

```diff
@@ -360,21 +395,107 @@ func NewVolumeOptionsFromVolID(
 		}
 	}
 
+	if imageAttributes.BackingSnapshotID != "" || volOptions.BackingSnapshotID != "" {
+		volOptions.BackingSnapshot = true
+		volOptions.BackingSnapshotID = imageAttributes.BackingSnapshotID
+	}
+
 	volOptions.ProvisionVolume = true
 	volOptions.SubVolume.VolID = vid.FsSubvolName
-	vol := core.NewSubVolume(volOptions.conn, &volOptions.SubVolume, volOptions.ClusterID)
+
+	if volOptions.BackingSnapshot {
+		err = volOptions.populateVolumeOptionsFromBackingSnapshot(ctx, cr)
+	} else {
+		err = volOptions.populateVolumeOptionsFromSubvolume(ctx)
+	}
+
+	return &volOptions, &vid, err
+}
+
+func (vo *VolumeOptions) populateVolumeOptionsFromSubvolume(ctx context.Context) error {
+	vol := core.NewSubVolume(vo.conn, &vo.SubVolume, vo.ClusterID)
+
+	var info *core.Subvolume
 	info, err := vol.GetSubVolumeInfo(ctx)
 	if err == nil {
-		volOptions.RootPath = info.Path
-		volOptions.Features = info.Features
-		volOptions.Size = info.BytesQuota
+		vo.RootPath = info.Path
+		vo.Features = info.Features
+		vo.Size = info.BytesQuota
 	}
 
 	if errors.Is(err, cerrors.ErrInvalidCommand) {
-		volOptions.RootPath, err = vol.GetVolumeRootPathCeph(ctx)
+		vo.RootPath, err = vol.GetVolumeRootPathCeph(ctx)
 	}
 
-	return &volOptions, &vid, err
+	return err
+}
+
+func (vo *VolumeOptions) populateVolumeOptionsFromBackingSnapshot(
+	ctx context.Context,
+	cr *util.Credentials,
+) error {
+	// As of CephFS snapshot v2 API, snapshots may be found in two locations:
+	//
+	// (a) /volumes/<volume group>/<subvolume>/.snap/<snapshot>/<UUID>
+	// (b) /volumes/<volume group>/<subvolume>/<UUID>/.snap/_<snapshot>_<snapshot inode number>
+
+	if !vo.ProvisionVolume {
+		// Case (b)
+		//
+		// If the volume is not provisioned by us, we assume that we have access only
```

Comment:
we are expecting to see snapshots created by admin manually in this case ? there could be other consumers or client who create snapshots however by making use of csi driver.  I am trying to understand which cases are expected to land here.

Reply:
Yes, that's precisely it -- pre-provisioned (static) subvolume snapshots. The use cases are the ones you've mentioned, as well as e.g. OpenStack Manila and other services where Ceph storage is managed by some external component.

I tried to briefly describe it in the docs https://github.com/ceph/ceph-csi/blob/f428a248c19995e7d90ce8d5f25f3991420344bf/docs/cephfs-snapshot-backed-volumes.md#mounting-snapshots-from-pre-provisioned-volumes, though I think there's some room for improvement to describe it better...

Reply:
Just FYI, speaking of pre-provisioned snapshots, there's a bug in latest ceph mds code that makes it impossible to access `<subvol root>/.snap/_<snap name>_<inode num>` if the snapshot name is too long (e.g. names generated by Manila). My colleague reported it here https://lists.ceph.io/hyperkitty/list/dev@ceph.io/thread/VREULFBGVN43TN4L3CR6LFXWN3NGJ7K3/ but it seems it has been known for some time and there are patches for it already available.